More Abstraction and Additional Clock Test Cases

juliand665 opened the pull request:
• I moved json-based result evaluation to a new protocol and made
BulletMatch and Clock use it.
• I generated more test cases for the clock challenge and made it use
json (see above). I believe these are correct, but I’m open to
criticism.

juliand665 committed:
Merge remote-tracking branch 'iosdevelopershq/master'

juliand665 committed:
More Abstraction and Additional Clock Test Cases

• I moved json-based result evaluation to a new protocol and made
BulletMatch and Clock use it.
• I generated more test cases for the clock challenge and made it use
json (see above). I believe these are correct, but I’m open to
criticism.

IanKeen commented on CodeChallenge/Base/CodeChallengeType.swift:
```diff
@@ -27,17 +29,67 @@ protocol CodeChallengeType {
+    
+    func generateDataset() -> [InputType] {
+        // Extract data from JSON file
+        let bundle = Bundle(for: BulletChallenge.self)
```
This is just to get the 'current' bundle correct?
If so it might be better to add a private junk class to use in place of `BulletChallenge` here so this doesn't mysteriously break in the future?

juliand665 replied:
Oh, whoops. I just copy-pasted that code from the BulletMatch code. Will fix.

juliand665 committed:
Corrected use of Bundle(for:)

missed this one when copy-pasting code from BulletMatch

juliand665 committed:
Fixed Indentation, Added Documentation

Trailing commas break auto-indentation :c
Also did some (very) minor cleanup

aranasaurus commented:
I'll leave it to @bugkrusha to merge since it is modifying his challenges a bit. But looks good to me! Thanks for the contribution!

juliand665 committed:
Changed closures to functions

Because apparently autocomplete doesn’t like closures here. This works
well for me

aranasaurus commented:
oh poop, I just merged another PR 'cause I thought this was already merged... gimme a min and I'll fix the conflicts

aranasaurus committed:
Merge branch 'master' into master
